torchx/schedulers/kueue_job_scheduler.py

```diff
@@ -0,0 +1,809 @@
+LABEL_APP_NAME = "torchx.pytorch.org/app-name"
+LABEL_ROLE_INDEX = "torchx.pytorch.org/role-index"
+LABEL_ROLE_NAME = "torchx.pytorch.org/role-name"
+LABEL_REPLICA_ID = "torchx.pytorch.org/replica-id"
+LABEL_KUBE_APP_NAME = "app.kubernetes.io/name"
+LABEL_ORGANIZATION = "app.kubernetes.io/managed-by"
+LABEL_UNIQUE_NAME = "app.kubernetes.io/instance"
+
+# Local Kueue and Priority class labels
+LOCAL_KUEUE_LABEL = "kueue.x-k8s.io/queue-name"
+PRIORITY_CLASS_LABEL = "kueue.x-k8s.io/priority-class"
+
+ANNOTATION_ISTIO_SIDECAR = "sidecar.istio.io/inject"
+
+LABEL_INSTANCE_TYPE = "node.kubernetes.io/instance-type"
+
+
+def sanitize_for_serialization(obj: object) -> object:
+    from kubernetes import client
+
+    api = client.ApiClient()
+    return api.sanitize_for_serialization(obj)
+
+
+def role_to_pod(name: str, role: Role, service_account: Optional[str]) -> "V1Pod":
+    from kubernetes.client.models import (  # noqa: F811 redefinition of unused
+        V1Container,
+        V1ContainerPort,
+        V1EmptyDirVolumeSource,
+        V1EnvVar,
+        V1HostPathVolumeSource,
+        V1ObjectMeta,
+        V1PersistentVolumeClaimVolumeSource,
+        V1Pod,
+        V1PodSpec,
+        V1ResourceRequirements,
+        V1SecurityContext,
+        V1Volume,
+        V1VolumeMount,
+    )
+
+    # limits puts an upper cap on the resources a pod may consume.
+    # requests is how much the scheduler allocates. We assume that the jobs will
+    # be allocation the whole machine so requests is slightly lower than the
+    # requested resources to account for the Kubernetes node reserved resources.
+    limits = {}
+    requests = {}
+
+    resource = role.resource
+    if resource.cpu > 0:
+        mcpu = int(resource.cpu * 1000)
+        limits["cpu"] = f"{mcpu}m"
+        request_mcpu = max(mcpu - RESERVED_MILLICPU, 0)
+        requests["cpu"] = f"{request_mcpu}m"
+    if resource.memMB > 0:
+        limits["memory"] = f"{int(resource.memMB)}M"
+        request_memMB = max(int(resource.memMB) - RESERVED_MEMMB, 0)
+        requests["memory"] = f"{request_memMB}M"
+    if resource.gpu > 0:
+        requests["nvidia.com/gpu"] = limits["nvidia.com/gpu"] = str(resource.gpu)
+
+    for device_name, device_limit in resource.devices.items():
+        limits[device_name] = str(device_limit)
+
+    resources = V1ResourceRequirements(
+        limits=limits,
+        requests=requests,
+    )
+
+    node_selector: Dict[str, str] = {}
+    if LABEL_INSTANCE_TYPE in resource.capabilities:
+        node_selector[LABEL_INSTANCE_TYPE] = resource.capabilities[LABEL_INSTANCE_TYPE]
+
+    # To support PyTorch dataloaders we need to set /dev/shm to larger than the
+    # 64M default so we mount an unlimited sized tmpfs directory on it.
+    SHM_VOL = "dshm"
+    volumes = [
+        V1Volume(
+            name=SHM_VOL,
+            empty_dir=V1EmptyDirVolumeSource(
+                medium="Memory",
+            ),
+        ),
+    ]
```

Comment:
Should we be worried about limiting the size of the shared memory dir? By default it can use up to 50% of the total allocated memory for the pod.